Lib/email/feedparser.py

```diff
@@ -87,40 +87,261 @@ def readline(self):
+            # If we had some sort of "StringViewIO" to avoid the copy, this
+            # would be significantly more efficient
+            # This code block, and the "else" code block below, functionally do
+            # the exact same thing, except this path makes no attempt to handle
+            # dumping data
+            sio = StringIO(data, '')
+            sio.seek(data_start_index)
+            lines = sio.readlines()
+            if lines:
+                if data[-1] != '\n':
+                    self._partial.append(lines.pop())
+                    if data[-1] == '\r':
+                        self._dangling_partial = True
+
+                self.pushlines(lines)
+        else:
+            dump_data_start = None if self._dump_destination is None \
+                                   else data_start_index
+            while unl_start_index >= 0:
+                unl_end_index = BufferedSubFile._find_unl_end(
+                    data, unl_start_index)
+                if unl_end_index < 0:
+                    # Incomplete line ending; break to just update our partial
+                    self._dangling_partial = True
+                    break
+
+                # We have an easy line; push it
+                if self._dump_destination is not None:
+                    # We have a window into a line. Make sure it's not EOF
+                    if self._check_eofstack(
+                        data, data_start_index, unl_end_index):
+                        # This line is "EOF". This is the end of our dump data
+                        self._dump_destination.append(
+                            data[dump_data_start:data_start_index])
+
+                        # Also push our line, since we already have it
+                        self._lines.append(
+                            data[data_start_index:unl_end_index])
+                        self._dump_destination = None
+                    #else: # This line didn't mark the end. Keep going.
+                else:
+                    # We're not dumping. Just go ahead and push the line
+                    self._lines.append(data[data_start_index:unl_end_index])
+
+                # Update our iterators
+                data_start_index = unl_end_index
+                unl_start_index = BufferedSubFile._find_unl(
+                    data, data_start_index)
+
+            if self._dump_destination is not None:
+                # Push everything that isn't going into the partial to the dump
+                # If we're able to safely flush the partial, do that too
+                # We don't care about self._is_dump_midline() here, because
+                # data_start_index always represents the start of a new line
+                if self._can_dump_partial(data, data_start_index):
+                    self._dump_destination.append(data[dump_data_start:])
+
+                    # Flush any partial-related state we may have set
```

Comment:
We're not really flushing the `self._partial` here as we don't do `self._partial.clear()` so it's a bit confusing.

Reply:
We may have set `_dangling_partial` on line 298 above. If we want to get rid of this, it'd be safer to replace `_dangling_partial` with a property or function to just check if the partial ends with the CR. My original goal with the flag was to track this state without explicitly checking every time. It's admittedly a bit confusing.